text/0000-dynamic-stylesheet-replacement.md

```diff
@@ -0,0 +1,260 @@
+import lightStylesheets from './light.css';
+
+export default StyleMixin(Base, lightStylesheets);
+```
+
+```html
+<!-- wrapper.html -->
+<template lwc:if={darkMode}>
+    <x-dark></x-dark>
+</template>
+<template lwc:else>
+    <x-light></x-light>
+</template>
+```
+
+```js
+// wrapper.js
+export default class extends LightningElement {
+  @api darkMode;
+}
+```
+
+</details>
+
+This is a lot of code! Plus it requires the additional wrapper component.
+
+Another goal of this RFC is to fully remove the [hacks that component authors are using](https://github.com/salesforce/lwc/issues/2826) to apply different stylesheets to the same template:
+
+```js
+import template from './template.html';
+import darkStylesheets from './dark.css';
+import lightStylesheets from './dark.css';
+
+export default class extends LightningElement {
+  @api darkMode;
+  
+  render() {
+    // DEPRECATED - DO NOT DO THIS!
+    template.stylesheets = darkMode ? darkStylesheets : lightStylesheets;
+    template.stylesheetToken = darkMode ? 'dark' : 'light';
+    return template;
+  }
+}
+```
+
+Programmatic stylesheets can replace some of these hacks, but not all. This proposal aims to replace _all_ of these use cases.
+
+## Detailed design
+
+LWC components are already able to [render multiple templates](https://lwc.dev/guide/html_templates#render-multiple-templates). However, the assumption is that component authors will
+only switch between different HTML files:
+
+```js
+import a from './a.html';
+import b from './b.html';
+
+export default class extends LightningElement {
+  @api shouldUseA;
+  
+  render() {
+    return shouldUseA ? a : b;
+  }
+}
+```
+
+(These HTML files may have their own implicitly-associated stylesheets, but component authors cannot reuse the same HTML file with different stylesheets.)
+
+This proposal augments the current behavior of the `render()` function. In addition to being able to return a template, the `render()` function can also return an object containing the `template` and `stylesheets` properties:
+
+```js
+import template from './template.html';
+import a from './a.css';
+import b from './b.css';
+
+export default class extends LightningElement {
+  @api shouldUseA;
+  
+  render() {
+    return {
+      template,
+      stylesheets: shouldUseA ? a : b
+    }
+  }
+}
+```
+
+The stylesheets are injected in the following order:
+
+1. Stylesheets implicitly associated with the `template`
+2. Stylesheets from the returned object's `stylesheets`
+3. Static `stylesheets` on the component
+
+This allows component authors to override the stylesheets that are implicitly associated with the `template`. For instance, they could have some shared default styles defined in `template.css`, whereas `a.css` and `b.css` would selectively override those defaults.
+
+LWC still makes no attempt to remove stylesheets, or to ensure a particular ordering _between_ `render()` calls. (This is in line with existing behavior.)

```

Comment:
Just for my understanding, what is the "ordering between render() calls" here referring to?

Reply:
I should be more specific. I mean ordering of injected stylesheets w.r.t the order that `render()` was called. I'll update this.